chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+The last requirement is achieved by using a stable sort during the post-processing.
+
+The algorithm used is listed in [@lst:nls-resolve-at].
+Given a concrete position, that is a `FileId` and `ByteIndex` in that file, a binary search is used to find the *last* element that *starts* at the given position.
+According to the aforementioned preconditions an element found there is equivalent to being the most specific element starting at this position.
+In the more frequent case that no element starting at the provided position is found, the search instead yields an index which can be used as a starting point to iterate the linearization *backwards* to find an item with the shortest span containing the queried position.
+Due to the third requirement, this reverse iteration can be aborted once an item's span ends before the query.
+If the search has to be aborted, the query does not have a corresponding `LinearizationItem`.
+
+```{.rust #lst:nls-resolve-at caption="Resolution of item at given position"}
+impl Completed {
+  pub fn item_at(
+    &self,
+    locator: &(FileId, ByteIndex),
+  ) -> Option<&LinearizationItem<Resolved>> {
+    let (file_id, start) = locator;
+    let linearization = &self.linearization;
+    let item = match linearization
+      .binary_search_by_key(
+        locator,
+        |item| (item.pos.src_id, item.pos.start))
+      {
+        // Found item(s) starting at `locator`
+        // search for most precise element
+        Ok(index) => linearization[index..]
+          .iter()
+          .take_while(|item| (item.pos.src_id, item.pos.start) == locator)
+          .last(),
+        // No perfect match found
+        // iterate back finding the first wrapping linearization item
+        Err(index) => {
+          linearization[..index].iter().rfind(|item| {
+            // Return the first (innermost) matching item
+            file_id == &item.pos.src_id
+            && start > &item.pos.start
+            && start < &item.pos.end
+          })
+        }
+    };
+    item
+  }
+}
+```
+
+#### Resolving by ID
+
+During the building process item IDs are equal to their index in the underlying List which allows for efficient access by ID.
+To allow similarly efficient access to nodes with using IDs a `Completed` linearization maintains a mapping of IDs to their corresponding index in the reordered array.
+A queried ID is first looked up in this mapping which yields an index from which the actual item is read.
+
+#### Resolving by scope
+
+During the construction from the AST, the syntactic scope of each element is eventually known.
+This allows to map scopes to a list of elements defined in this scope.
+Definitions from higher scopes are not repeated, instead they are calculated on request.
+As scopes are lists of scope fragments, for any given scope the set of referable nodes is determined by unifying IDs of all prefixes of the given scope, then resolving the IDs to elements.
```

Comment:
It may be worth it to explain just slightly more how are scopes represented in practice. Given this sentences, it seems like they are linked lists of arrays, very similar to the `nickel-lang::environment::Environment` data structure, that are then flattened when requesting an identifier?

Reply:
Its a Map of arrays actually and not too dissimilar from the layers of `Envrionment`
The differnce is that layers are represented in a flat structure and persistent.
Each layer (if you will) is uniquely identified by its scope(-path), the elements of which are are the prefix I refer to.

Without thinking of a concrete Nickel example the scope could look like this:

```
1 -> [Node(1)]
1/1 -> [Node(2), Node(6)]
1/2 -> [Node(3)],
1/2/1 -> [Node(4)],
1/2/2 -> [Node(5)]
```

If we resolve the names in scope for `Node(5)` we get

1 ->  [Node(1)]
1/2 -> [Node(3)]

=> [Node(1), Node(3)]
 
In fact this should be the same as if we stored the environment for every node, just more efficient(?)

Reply:
@yannham 
Changed this to [15ea55c9cd2c743918e35e5734ef6cc47f52f67a](https://github.com/ysndr/MSc-Thesis/pull/4/commits/15ea55c9cd2c743918e35e5734ef6cc47f52f67a)

Do you think this makes it clearer?

Reply:
Yes, the diagram in particular makes things clearer.